lib/graphql/query/serial_execution/selection_resolution.rb

```diff
@@ -19,34 +13,94 @@ def initialize(target, type, selections, query, execution_strategy)
 
-        def deep_merge(h1, h2)
-          h1.merge(h2) do |key, oldval, newval|
-            if oldval.is_a?(Array) && newval.is_a?(Array)
-              oldval.each_index.map do |i|
-                deep_merge oldval[i], newval[i]
-              end
-            elsif oldval.is_a?(Hash) && newval.is_a?(Hash)
-              deep_merge(oldval, newval)
-            else
-              newval
+        def flatten_selection(ast_node)
+          return {(ast_node.alias || ast_node.name) => ast_node} if ast_node.is_a?(GraphQL::Language::Nodes::Field)
+
+          ast_fragment = get_fragment(ast_node)
+          return {} unless fragment_type_can_apply?(ast_fragment)
+
+          chain = GraphQL::Query::DirectiveChain.new(ast_node, query) {
+            ast_fragment.selections.reduce({}) do |memo, selection|
+              memo.merge(flatten_selection(selection))
             end
+          }
+
+          chain.result || {}
+        end
+
+        def get_fragment(ast_node)
+          if ast_node.is_a? GraphQL::Language::Nodes::FragmentSpread
+            query.fragments[ast_node.name]
+          elsif ast_node.is_a? GraphQL::Language::Nodes::InlineFragment
+            ast_node
+          else
+            raise 'Unrecognized fragment node'
           end
         end
 
-        def resolve_field(ast_field)
-          chain = GraphQL::Query::DirectiveChain.new(ast_field, query) {
-            strategy_name = RESOLUTION_STRATEGIES[ast_field.class]
-            strategy_class = execution_strategy.public_send(strategy_name)
-            strategy = strategy_class.new(ast_field, type, target, query, execution_strategy)
-            strategy.result
+        def fragment_type_can_apply?(ast_fragment)
+          child_type = query.schema.types[ast_fragment.type]
+          resolved_type = GraphQL::Query::TypeResolver.new(target, child_type, type).type
+          !resolved_type.nil?
+        end
+
+        def merge_fields(field1, field2)
+          if fields_can_merge?(field1, field2)
+            # create a new ast field node merging selections from each field
+            GraphQL::Language::Nodes::Field.new(
+              name: field2.name,
+              alias: field2.alias,
+              arguments: field2.arguments,
+              directives: field1.directives + field2.directives,
+              selections: field1.selections + field2.selections
+            )
+          else
+            field2
+          end
+        end
+
+        def fields_can_merge?(field1, field2)
+          type1 = query.schema.get_field(type, field1.name).type
+          type2 = query.schema.get_field(type, field2.name).type
+
+          while type1.is_a?(GraphQL::ListType) || type1.is_a?(GraphQL::NonNullType)
+            type1 = type1.of_type
+          end
+
+          while type2.is_a?(GraphQL::ListType) || type2.is_a?(GraphQL::NonNullType)
+            type2 = type2.of_type
+          end
+
+          type1.is_a?(GraphQL::ObjectType) && type2.is_a?(GraphQL::ObjectType)
```

Comment:
How come we only have to check that they're both ObjectTypes? Shouldn't we also check that they're the _same_ object type? 

Oh, I guess you know they're basically the same already, since you're getting fields with the same name from the same type. 

Ok, I think I get it :)